File: app/src/main/java/com/pawoon/test/presenter/connection/service/ApiClient.java
package com.pawoon.test.presenter.connection.service;

import android.app.Activity;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.pawoon.test.BuildConfig;
import com.pawoon.test.presenter.base.session.SessionToken;

import java.io.IOException;
import java.util.concurrent.TimeUnit;

import okhttp3.Interceptor;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.Response;
import retrofit2.Retrofit;
import retrofit2.adapter.rxjava2.RxJava2CallAdapterFactory;
import retrofit2.converter.gson.GsonConverterFactory;

public class ApiClient extends Activity {
    public static final String BASE_URL = BuildConfig.URL_BASE;
    private static Retrofit retrofit = null;


    public static Retrofit getClient(Activity activity) {
        final SessionToken token = new SessionToken(activity);
        Interceptor interceptor = new Interceptor() {
            @Override
            public Response intercept(Chain chain) throws IOException {
                final Request request = chain.request().newBuilder()
                        .addHeader("X-AUTH-TOKEN", token.getToken())
                        .build();
                return chain.proceed(request);
            }
        };
        OkHttpClient.Builder httpClient = new OkHttpClient.Builder();
        httpClient.addInterceptor(interceptor);
        httpClient.connectTimeout(5, TimeUnit.MINUTES);
        httpClient.readTimeout(5, TimeUnit.MINUTES);
        OkHttpClient client = httpClient.build();

        RxJava2CallAdapterFactory rxAdapter = RxJava2CallAdapterFactory.create();


        Gson gson = new GsonBuilder()
                .setLenient()
                .create();

        if (retrofit==null) {
            retrofit = new Retrofit.Builder()
                    .baseUrl(BASE_URL)
                    .addConverterFactory(GsonConverterFactory.create(gson))
                    .addCallAdapterFactory(rxAdapter)
                    .client(client)
                    .build();
        }
        return retrofit;
    }
}

File: app/src/main/java/com/project/base/presenter/connection/service/ApiClient.kt
package com.project.base.presenter.connection.service

import android.app.Activity


import com.project.base.BuildConfig
import com.project.base.presenter.base.session.SessionToken

import java.io.IOException
import java.util.concurrent.TimeUnit

import okhttp3.Interceptor
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.Response
import retrofit2.Retrofit
import retrofit2.adapter.rxjava.RxJavaCallAdapterFactory
import retrofit2.converter.gson.GsonConverterFactory

/**
 * Created by faizf on 2/8/2017.
 */

class ApiClient : Activity() {
    companion object {

        val BASE_URL = BuildConfig.URL_BASE
        private var retrofit: Retrofit? = null


        fun getClient(activity: Activity): Retrofit {
            val token = SessionToken(activity)
            val interceptor = Interceptor { chain ->
                val request = chain.request().newBuilder()
                        .addHeader("X-AUTH-TOKEN", token.token)
                        .build()
                chain.proceed(request)
            }
            val httpClient = OkHttpClient.Builder()
            httpClient.addInterceptor(interceptor)
            httpClient.connectTimeout(5, TimeUnit.MINUTES)
            httpClient.readTimeout(5, TimeUnit.MINUTES)
            val client = httpClient.build()

            val rxAdapter = RxJavaCallAdapterFactory.create()

            if (retrofit == null) {
                retrofit = Retrofit.Builder()
                        .baseUrl(BASE_URL)
                        .addConverterFactory(GsonConverterFactory.create())
                        .addCallAdapterFactory(rxAdapter)
                        .client(client)
                        .build()
            }
            return retrofit
        }
    }
}
